google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/GapicSpannerRpc.java

```diff
@@ -491,6 +504,73 @@ public GapicSpannerRpc(final SpannerOptions options) {
     }
   }
 
+  private static String parseGrpcGcpApiConfig() {
+    InputStream inputStream = GapicSpannerRpc.class.getResourceAsStream(API_FILE);
+    StringBuilder sb = new StringBuilder();
+    try {
+      for (int ch; (ch = inputStream.read()) != -1; ) {
+        sb.append((char) ch);
+      }
+    } catch (IOException e) {
+      throw newSpannerException(e);
+    }
+    return sb.toString();
+  }
+
+  // Enhance metric options for gRPC-GCP extension. Adds metric registry if not specified. Adds
+  // more labels on top of provided.
+  private static GcpManagedChannelOptions grpcGcpOptionsWithMetrics(SpannerOptions options) {
+    GcpManagedChannelOptions grpcGcpOptions =
+        MoreObjects.firstNonNull(options.getGrpcGcpOptions(), new GcpManagedChannelOptions());
+    GcpMetricsOptions metricsOptions = MoreObjects.firstNonNull(
+        grpcGcpOptions.getMetricsOptions(), GcpMetricsOptions.newBuilder().build());
+    GcpMetricsOptions.Builder metricsOptionsBuilder = GcpMetricsOptions.newBuilder(metricsOptions);
+    if (metricsOptions.getMetricRegistry() == null) {
+      metricsOptionsBuilder.withMetricRegistry(Metrics.getMetricRegistry());
+    }
+    List<LabelKey> labelKeys = metricsOptions.getLabelKeys();
+    List<LabelValue> labelValues = metricsOptions.getLabelValues();
+    // TODO: Add default labels with values: client_id, database, instance_id.
```

Comment:
I see you would like to get label values for client_id, instance_id and database_id within a static method, which only has access SpannerOptions. SpannerOptions is not tied directly one database / instance, so you won't be able to do this here.

For the client id (which I imagine is the client language + version) you can do the following: https://github.com/googleapis/java-spanner/blob/6ce4c2da0e1378011793fd1fa136f669624234a1/google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/GapicSpannerRpc.java#L250-L251

In terms of the resource we are using for a specific request, that is tied up to the session which is being used. We do add an extra header in every call context as specified [here](https://github.com/googleapis/java-spanner/blob/6ce4c2da0e1378011793fd1fa136f669624234a1/google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/GapicSpannerRpc.java#L1630). The extra header key name will be [google-cloud-resource-prefix](https://github.com/googleapis/gax-java/blob/49efdc38f2d6ca3838dcecee63dc6cd07f6c2b23/gax/src/main/java/com/google/api/gax/rpc/ApiClientHeaderProvider.java#L99) and the value will be one of [these two](https://github.com/googleapis/java-spanner/blob/47f59655acaf0fcf5848039e6b4881b80cdceb5c/google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/SpannerMetadataProvider.java#L32-L35) depending on whether you are performing an operation on an instance only resource or an instance + database resource.